src/server/saved_objects/migrations/core/diff_mapping.ts

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import _ from 'lodash';
+import { IndexMapping } from './call_cluster';
+
+export enum MigrationAction {
+  None = 0,
+  Patch = 1,
+  Migrate = 2,
+}
+
+/**
+ * Provides logic that diffs the actual index mappings with the expected
+ * mappings. It ignores differences in dynamic mappings.
+ *
+ * If mappings differ in a patchable way, the result is 'patch', if mappings
+ * differ in a way that requires migration, the result is 'migrate', and if
+ * the mappings are equivalent, the result is 'none'.
+ */
+export function diffMapping(actual: IndexMapping, expected: IndexMapping): MigrationAction {
+  if (actual.doc.dynamic !== expected.doc.dynamic) {
+    return MigrationAction.Migrate;
+  }
+
+  const actualProps = actual.doc.properties;
```

Comment:
The more I think about this - we could make these changes now.

Reply:
Interesting. Are there potentially negative consequences to doing this in 6.x?

Reply:
I don't believe so - once we introduce migrations we will immediately migrate the index and create the alias. My immediate thought is that this will be easier to do during this period, rather than changing the types during 6.6. Thoughts?

Reply:
I don't disagree with the general sentiment here, but let's not add any actual migrations to this PR. Not only is this PR hard enough to grok, but if we did need to revert a migration because it turned out to have broader impacts than we anticipated, it'd be much easier to do if it was contained in its own PR.

Reply:
We chatted about this - the `doc` to `_doc` change will touch a lot. We're going to focus on getting this in and will worry about that after. It's possible that effort will be done in the inital ES re-index when we create the alias. This can also happen in 7.0.0.